packages/mui-material-next/src/FilledInput/FilledInput.tsx

```diff
@@ -145,115 +155,174 @@ const FilledInputInput = styled(InputBaseInput, {
+              '&:-webkit-autofill': {
+                WebkitBoxShadow: '0 0 0 100px #266798 inset',
+                WebkitTextFillColor: '#fff',
+                caretColor: '#fff',
+              },
+            },
+          }),
+        }),
+    ...(ownerState.size === 'small' && {
+      paddingTop: 21,
+      paddingBottom: 4,
+    }),
+    ...(ownerState.hiddenLabel && {
+      paddingTop: 16,
+      paddingBottom: 17,
+    }),
+    ...(ownerState.multiline && {
+      paddingTop: 0,
+      paddingBottom: 0,
+      paddingLeft: 0,
+      paddingRight: 0,
+    }),
+    ...(ownerState.startAdornment && {
+      paddingLeft: 0,
+    }),
+    ...(ownerState.endAdornment && {
+      paddingRight: 0,
     }),
-}));
+    ...(ownerState.hiddenLabel &&
+      ownerState.size === 'small' && {
+        paddingTop: 8,
+        paddingBottom: 9,
+      }),
+  };
+});
 
-const FilledInput = React.forwardRef(function FilledInput(inProps, ref) {
+const FilledInput = React.forwardRef(function FilledInput<
+  RootComponentType extends React.ElementType,
+>(inProps: FilledInputProps<RootComponentType>, forwardedRef: React.ForwardedRef<Element>) {
   const props = useThemeProps({ props: inProps, name: 'MuiFilledInput' });
 
   const {
     disableUnderline,
-    components = {},
-    componentsProps: componentsPropsProp,
     fullWidth = false,
     hiddenLabel, // declare here to prevent spreading to DOM
     inputComponent = 'input',
     multiline = false,
-    slotProps,
-    slots = {},
     type = 'text',
+    slotProps = {},
+    slots = {},
     ...other
   } = props;
 
-  const ownerState = {
+  const ownerState: FilledInputOwnerState = {
     ...props,
+    disableUnderline,
     fullWidth,
     inputComponent,
     multiline,
     type,
   };
 
-  const classes = useUtilityClasses(props);
-  const filledInputComponentsProps = { root: { ownerState }, input: { ownerState } };
+  const classes = useUtilityClasses(ownerState);
+
+  const Root = slots.root ?? FilledInputRoot;
+  const Input = slots.input ?? FilledInputInput;
+
+  const rootProps = useSlotProps({
+    elementType: Root,
+    externalSlotProps: slotProps.root,
+    additionalProps: {
+      ref: forwardedRef,
+      fullWidth,
+      inputComponent,
+    },
+    externalForwardedProps: other,
+    ownerState: ownerState as FilledInputOwnerState & InputBaseOwnerState,
+    className: [classes.root],
+  });
 
-  const componentsProps =
-    slotProps ?? componentsPropsProp
-      ? deepmerge(slotProps ?? componentsPropsProp, filledInputComponentsProps)
-      : filledInputComponentsProps;
+  const inputProps = useSlotProps({
+    elementType: Input,
+    externalSlotProps: slotProps.input,
+    ownerState: ownerState as FilledInputOwnerState & InputBaseOwnerState,
+    className: [classes.input],
+  });
 
-  const RootSlot = slots.root ?? components.Root ?? FilledInputRoot;
-  const InputSlot = slots.input ?? components.Input ?? FilledInputInput;
+  if (multiline) {
```

Comment:
This is for TS - it's very fussy about `multiline` and `type`, I couldn't pass them through `useSlotProps` without a ton of TS errors 😓

Reply:
What was the error? 🤔

Reply:
Something messy 😓 
```sh
FilledInput.tsx:283:6 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: { inputComponent?: "symbol" | "object" | "search" | "big" | "link" | "small" | "sub" | "sup" | "map" | "filter" | "input" | "base" | "time" | "code" | "data" | "progress" | "stop" | ... 162 more ... | undefined; } & OverrideProps<...>): Element | null', gave the following error.
    Type '{ ref: ((MutableRefObject<Element | null> | ((instance: Element | null) => void)) & ((instance: any) => void)) | null; fullWidth: boolean; inputComponent: ElementType<...>; ... 6 more ...; type: HTMLInputTypeAttribute; }' is not assignable to type 'IntrinsicAttributes & ({ inputComponent?: "symbol" | "object" | "search" | "big" | "link" | "small" | "sub" | "sup" | "map" | "filter" | "input" | "base" | ... 167 more ... | undefined; } & OverrideProps<...>)'.
      Types of property 'multiline' are incompatible.
        Type 'boolean' is not assignable to type 'false'.
```

However, if we don't do `InputBaseOwnProps = (SingleLineInputProps | MultiLineInputProps) & ...` (i.e. https://github.com/mui/material-ui/pull/39307/commits/227a2b310e7b40b7ad3acac6f4ca89c3ad67739b) it seems fine 🤔 

Maybe I overlooked something? Wonder if @michaldudak has some insight on this, I originally based `SingleLineInputProps | MultiLineInputProps` on the equivalent types in Base UI's `Input.types`

Reply:
```jsx
<InputBase
  slots={{ root: Root, input: Input }}
  slotProps={{
    input: inputProps,
  }}
  type={type}
  multiline={multiline}
  {...rootProps}
/>
```

The error you see comes from the fact you're providing the `type` prop, which makes TS choose the overload with `multiline=false`. And since the `multiline` variable can be `true | false`, TS complains.

It would be nice if we could do
```
type={multiline ? undefined : type}
multiline={multiline}
```

but TS isn't that smart, apparently :)

Reply:
> If we don't do InputBaseOwnProps = (SingleLineInputProps | MultiLineInputProps) & ... (i.e. https://github.com/mui/material-ui/commit/227a2b310e7b40b7ad3acac6f4ca89c3ad67739b) and kept it like before everything seems fine 🤔

What do you think it's best? I would prefer not to have the `if (multiline) ...`, unless the types become too messy—no hard preference from my side.

Reply:
Let's keep `InputBaseOwnProps` flat like it was before then, in this case I think its obvious enough the several multiline related props are only relevant for a multiline input 😬

Reply:
> It would be nice if we could do
> 
> ```
> type={multiline ? undefined : type}
> multiline={multiline}
> ```
> 
> but TS isn't that smart, apparently :)

@michaldudak I see ~ I did experiment a bit with `undefined` and conditions, wasn't sure if there were any weird tricks or workarounds for it 😆